pkg/assembler/backends/testing/hasSourceAt.go

```diff
@@ -112,92 +233,46 @@ func (c *demoClient) registerHasSourceAt(selectedPackage *model.Package, selecte
 	return newHasSourceAt, nil
 }
 
-func (c *demoClient) IngestHasSourceAt(ctx context.Context, pkg model.PkgInputSpec, pkgMatchType model.MatchFlags, source model.SourceInputSpec, hasSourceAt model.HasSourceAtInputSpec) (*model.HasSourceAt, error) {
-
-	var selectedPkgSpec *model.PkgSpec
-	if pkgMatchType.Pkg == model.PkgMatchTypeSpecificVersion {
-		selectedPkgSpec = helper.ConvertPkgInputSpecToPkgSpec(&pkg)
-
-	} else {
-		selectedPkgSpec = &model.PkgSpec{
-			Type:      &pkg.Type,
-			Namespace: pkg.Namespace,
-			Name:      &pkg.Name,
-		}
-	}
-	collectedPkg, err := c.Packages(ctx, selectedPkgSpec)
-	if err != nil {
-		return nil, err
-	}
-	if len(collectedPkg) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestCertifyBad :: multiple packages found")
-	}
-
-	sourceSpec := helper.ConvertSrcInputSpecToSrcSpec(&source)
-
-	sources, err := c.Sources(ctx, sourceSpec)
-	if err != nil {
-		return nil, err
-	}
-	if len(sources) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestOccurrence :: source argument must match one"+
-				" single source repository, found %d",
-			len(sources))
-	}
-	return c.registerHasSourceAt(
-		collectedPkg[0],
-		sources[0],
-		hasSourceAt.KnownSince,
-		hasSourceAt.Justification,
-		hasSourceAt.Origin,
-		hasSourceAt.Collector)
-}
-
 // Query HasSourceAt
 
-func (c *demoClient) HasSourceAt(ctx context.Context, hasSourceAtSpec *model.HasSourceAtSpec) ([]*model.HasSourceAt, error) {
+func (c *demoClient) HasSourceAt(ctx context.Context, filter *model.HasSourceAtSpec) ([]*model.HasSourceAt, error) {
+	out := []*model.HasSourceAt{}
 
-	var collectedHasSourceAt []*model.HasSourceAt
-
-	for _, h := range c.hasSourceAt {
-		matchOrSkip := true
-
-		if hasSourceAtSpec.Justification != nil && h.Justification != *hasSourceAtSpec.Justification {
-			matchOrSkip = false
+	for _, mapLink := range sourceMaps {
+		if noMatch(filter.Justification, mapLink.justification) || noMatch(filter.Origin, mapLink.origin) || noMatch(filter.Collector, mapLink.collector) {
+			continue
 		}
-		if hasSourceAtSpec.Collector != nil && h.Collector != *hasSourceAtSpec.Collector {
-			matchOrSkip = false
+		// Note: Here we may call the selection resolvers since we
+		// build GraphQL structs based on what's on the backend. But
+		// this will result in then having to build a cartesian product
+		// between packages (of count, say, P) and sources (of count,
+		// say, C), for a total of P*C nodes than then need to be
+		// compared with the, say, M sourceMaps that we select so far.
+		// In general M << {P, C}, so this is wasteful.
```

Comment:
Very minor nit: We can remove this, that was first for me as a reminder to not do cartesian product